numpy/core/src/multiarray/descriptor.c

```diff
@@ -348,14 +350,21 @@ _convert_from_tuple(PyObject *obj)
             }
         }
         items = PyArray_OverflowMultiplyList(shape.ptr, shape.len);
-        if ((items < 0) || (items > (NPY_MAX_INT / type->elsize))) {
+        if (items < 0 || items > NPY_MAX_INT) {
+            overflowed = 1;
+        }
+        else {
+            overflowed = npy_mul_with_overflow_int(
+                &nbytes, type->elsize, (int) items);
+        }
+        if (overflowed) {
             PyErr_SetString(PyExc_ValueError,
                             "invalid shape in fixed-type tuple: dtype size in "
                             "bytes must fit into a C int.");
             npy_free_cache_dim_obj(shape);
             goto fail;
         }
-        newdescr->elsize = type->elsize * items;
+        newdescr->elsize = nbytes;
         if (newdescr->elsize == -1) {
```

Comment:
Can `nbytes` be -1? If so, could use a note. I'm guessing a single item of size 0, which seems restrictive.

Reply:
Huh, that's pretty weird. This couldn't happen before either, could it?